Add network attachment to gce instance

maxi-cit opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->
This PR enables support for network attachment at the gce instance network interfaces
<!--
Replace each [ ] with [X] to check it. Switch to the preview view to make it easier to click on links.
These steps will speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.
-->
If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [x] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/develop/run-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
google_compute_instance
```

modular-magician commented:
Hello! I am a robot. It looks like you are a: Community Contributor ~Googler~ ~Core Contributor~. Tests will require approval to run.

@trodge, a repository maintainer, has been assigned to review your changes. If you have not received review feedback within 2 business days, please leave a comment on this PR asking them to take a look.

You can help make sure that review is quick by [doing a self-review](https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/) and by [running impacted tests locally](https://googlecloudplatform.github.io/magic-modules/get-started/run-provider-tests/).

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-8829-old..auto-pr-8829) ( 2 files changed, 285 insertions(+))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_compute_instance_from_machine_image` (5 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_compute_instance_from_machine_image" "primary" {
  network_interface {
    network_attachment = # value needed
  }
}

```
Resource: `google_compute_instance_from_template` (10 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_compute_instance_from_template" "primary" {
  network_interface {
    network_attachment = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3010` 
Passed tests `2712` 
Skipped tests: `296` 
Affected tests: `2` 

#### Action taken 
<details> <summary>Found 2 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccComputeInstance_networkAttachmentUpdate|TestAccComputeInstance_NetworkAttachment </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccComputeInstance_networkAttachmentUpdate`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/3c082272-dd07-49a0-8616-ffa78c64943c/build-log/recording_build/TestAccComputeInstance_networkAttachmentUpdate_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/3c082272-dd07-49a0-8616-ffa78c64943c/recording/TestAccComputeInstance_networkAttachmentUpdate.log)]
`TestAccComputeInstance_NetworkAttachment`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/3c082272-dd07-49a0-8616-ffa78c64943c/build-log/recording_build/TestAccComputeInstance_NetworkAttachment_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/3c082272-dd07-49a0-8616-ffa78c64943c/recording/TestAccComputeInstance_NetworkAttachment.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/3c082272-dd07-49a0-8616-ffa78c64943c/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/3c082272-dd07-49a0-8616-ffa78c64943c/recording) for each test

trodge commented:
Tests are failing with the following error:
```
2023-09-02T00:21:13.173Z [ERROR] sdk.proto: Response contains error diagnostic: tf_provider_addr=registry.terraform.io/hashicorp/google tf_proto_version=5.3 diagnostic_detail="" diagnostic_severity=ERROR diagnostic_summary="Error creating instance: googleapi: Error 400: Invalid value for field 'resource.networkInterfaces[1]': '{  \"networkAttachment\": \"https://www.googleapis.com/compute/beta/projects/ci-test-project-188019/reg...'. Instance must be in the same region as the network attachment., invalid" tf_rpc=ApplyResourceChange tf_resource_type=google_compute_instance tf_req_id=1bac79aa-4c87-e3ed-6d10-8018caf72a57
```

trodge commented on mmv1/third_party/terraform/services/compute/resource_compute_instance_test.go.erb:
```diff
@@ -6829,3 +6909,188 @@ resource "google_compute_disk" "debian" {
+resource "google_compute_instance" "foobar" {
+	name         = "%{instance_name}"
+	machine_type = "e2-medium"
+	zone         = "us-east1-d"
```
I believe this needs to be a zone in us-central1.

trodge commented on mmv1/third_party/terraform/services/compute/resource_compute_instance_test.go.erb:
```diff
@@ -6829,3 +6909,188 @@ resource "google_compute_disk" "debian" {
+
+resource "google_compute_network_attachment" "test_network_attachment2" {
+	name = "tf-test-network-attachment2-%{suffix}"
+    region = "us-central1"
```
Please fix the indentation here.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-8829-old..auto-pr-8829) ( 1 file changed, 11 insertions(+), 8 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-8829-old..auto-pr-8829) ( 3 files changed, 316 insertions(+), 8 deletions(-))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_compute_instance_from_machine_image` (5 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_compute_instance_from_machine_image" "primary" {
  network_interface {
    network_attachment = # value needed
  }
}

```
Resource: `google_compute_instance_from_template` (10 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_compute_instance_from_template" "primary" {
  network_interface {
    network_attachment = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3072` 
Passed tests `2766` 
Skipped tests: `300` 
Affected tests: `6` 

#### Action taken 
<details> <summary>Found 6 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccBigQueryDataTable_bigtable|TestAccBigtableAppProfile_bigtableAppProfileAnyclusterExample|TestAccBigtableAppProfile_bigtableAppProfileSingleclusterExample|TestAccBigtableAppProfile_bigtableAppProfileMulticlusterExample|TestAccComputeInstance_networkAttachmentUpdate|TestAccComputeInstance_NetworkAttachment </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccBigQueryDataTable_bigtable`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/1f298962-1d97-4959-95b2-a8283f0ab5da/recording/TestAccBigQueryDataTable_bigtable.log)]
`TestAccBigtableAppProfile_bigtableAppProfileAnyclusterExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/1f298962-1d97-4959-95b2-a8283f0ab5da/recording/TestAccBigtableAppProfile_bigtableAppProfileAnyclusterExample.log)]
`TestAccBigtableAppProfile_bigtableAppProfileSingleclusterExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/1f298962-1d97-4959-95b2-a8283f0ab5da/recording/TestAccBigtableAppProfile_bigtableAppProfileSingleclusterExample.log)]
`TestAccBigtableAppProfile_bigtableAppProfileMulticlusterExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/1f298962-1d97-4959-95b2-a8283f0ab5da/recording/TestAccBigtableAppProfile_bigtableAppProfileMulticlusterExample.log)]
`TestAccComputeInstance_NetworkAttachment`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/1f298962-1d97-4959-95b2-a8283f0ab5da/recording/TestAccComputeInstance_NetworkAttachment.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccComputeInstance_networkAttachmentUpdate`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/1f298962-1d97-4959-95b2-a8283f0ab5da/build-log/recording_build/TestAccComputeInstance_networkAttachmentUpdate_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/1f298962-1d97-4959-95b2-a8283f0ab5da/recording/TestAccComputeInstance_networkAttachmentUpdate.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/1f298962-1d97-4959-95b2-a8283f0ab5da/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/1f298962-1d97-4959-95b2-a8283f0ab5da/recording) for each test

maxi-cit commented:
Hello @trodge I updated the behavior of this field since this [API](https://cloud.google.com/compute/docs/reference/rest/beta/instances/updateNetworkInterface) is not able to update an interface's network attachment. This makes this field immutable and will require instance recreation
I also updated the related tests

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-8829-old..auto-pr-8829) ( 1 file changed, 11 insertions(+), 8 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-8829-old..auto-pr-8829) ( 4 files changed, 228 insertions(+), 17 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-8829-old..auto-pr-8829) ( 1 file changed, 9 deletions(-))
## Missing test report
Your PR includes resource fields which are not covered by any test.

Resource: `google_compute_instance_from_machine_image` (5 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_compute_instance_from_machine_image" "primary" {
  network_interface {
    network_attachment = # value needed
  }
}

```
Resource: `google_compute_instance_from_template` (10 total tests)
Please add an acceptance test which includes these fields. The test should include the following:

```hcl
resource "google_compute_instance_from_template" "primary" {
  network_interface {
    network_attachment = # value needed
  }
}

```

modular-magician commented:
#### Tests analytics 
Total tests: `3077` 
Passed tests `2777` 
Skipped tests: `299` 
Affected tests: `1` 

#### Action taken 
<details> <summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccComputeInstance_NetworkAttachmentUpdate </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccComputeInstance_NetworkAttachmentUpdate`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/217a6459-f272-4abf-89bf-451c9b97c994/recording/TestAccComputeInstance_NetworkAttachmentUpdate.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{green}{\textsf{All tests passed!}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/217a6459-f272-4abf-89bf-451c9b97c994/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-8829/artifacts/217a6459-f272-4abf-89bf-451c9b97c994/recording) for each test
